Square root function

jacqueswww opened the pull request:
### What I did
Fixes #1266.

### How I did it

### How to verify it

### Description for the changelog

Add sqrt function.

### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://i.pinimg.com/originals/3d/ec/fe/3decfe84de9f436b3785f09919e6c701.jpg)

jacqueswww commented on vyper/functions/functions.py:
```diff
@@ -1224,6 +1227,61 @@ def _can_compare_with_uint256(operand):
+        break
+    y = z
+    z = (x / z + z) / 2.0
+    """
```
@charles-cooper I think this is beginning of our own virtual function implementation :)

charles-cooper replied:
Neat!

fubuloubu commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -0,0 +1,63 @@
+    return val.sqrt().quantize(
+        Decimal('0.0000000000'),
+        rounding=ROUND_FLOOR
+    )
```
Just to confirm, this will match up to 10 decimal places, correct?

jacqueswww replied:
Yes, took me a while to find it ;)

fubuloubu replied:
Make sure to test around the edge of this a little better (i.e. something at 9, 10, 11 decimal places)

jacqueswww replied:
Not sure what you mean? The round/truncate is applied after the square root ( which has way more digits than 10).

fubuloubu replied:
Oh, I mean that the test cases you run through contract produce results that have significant figures at 9, 10, and 11 decimals in length, to verify this truncation does not lose precision.

jacqueswww commented:
Got it, have a few other tests I want to add in anycase :)

davesque commented on vyper/parser/memory_allocator.py:
```diff
@@ -0,0 +1,23 @@
+
+
+class MemoryAllocator():
+
```
We should probably also add an instance property type annotation here:
```suggestion
next_mem: int

```

davesque commented on vyper/parser/memory_allocator.py:
```diff
@@ -0,0 +1,24 @@
+    def __init__(self, start_position: int = MemoryPositions.RESERVED_MEMORY):
+        self.next_mem = start_position
+
+    # Get the next unused memory location
```
Actually, one other thing.  I've noticed that we're often using comments above method and function definitions when it's more customary to use docstrings.  Can we switch to docstrings or is there a particular reason it's being done this away?

jacqueswww replied:
@davesque main reason is that was the convention on repo when I joined, also I think it reads well, especially because there is a difference between comments and docs. Comments augment the code or 'narrative' of the code were reading the code alone isn't good enough, docblocks is good for understanding how to call a function. The fact that docblocks are required to be fat by convention, supports the point. A good balance between docblocks and comments is what I prefer over a codebase with too much docblocks. But that's just my completely biased and super objective opinion.

charles-cooper commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -0,0 +1,63 @@
+)
+
+
+def decimal_sqrt(val):
```
I prefer to have a separate function for rounding
```suggestion
def decimal_round(val, places, rounding=ROUND_FLOOR):
  q = '0'
  if places != 0 : 
    q += '.' + '0' * places
  return val.quantize(Decimal(q), rounding=rounding)
```
That way it's easier to check at a glance the number of places being rounded

jacqueswww replied:
I think your example is missing the .sqrt() ;)

charles-cooper commented on vyper/parser/memory_allocator.py:
```diff
@@ -0,0 +1,24 @@
+    # Grow memory by x bytes
+    def increase_memory(self, size: int) -> Tuple[int, int]:
+        before_value = self.next_mem
+        self.next_mem += size
```
Now that this is encapsulated in a function, we can double check that it's 32 aligned.
```suggestion
        self.next_mem += size
        assert self.next_mem % 32 == 0
```

jacqueswww replied:
Good idea :+1:

pipermerriam replied:
We've tended to avoid using assertions in runtime code because they are easily disable.

https://github.com/ethereum/snake-charmers-tactical-manual/issues/22

charles-cooper replied:
Maybe we should add a `CompilerPanic` exception? This could also automate adding 'please file an issue' to the exception message.

jacqueswww replied:
@pipermerriam yip already did it with an exception ;)

charles-cooper commented on vyper/functions/functions.py:
```diff
@@ -1224,6 +1227,61 @@ def _can_compare_with_uint256(operand):
+    return LLLnode.from_list(
+        [
+            'seq_unchecked',
+            ['if', ['eq', arg, 0], 0],
```
I don't think this short-circuits?

charles-cooper replied:
It's probably easiest to write this into the inline sqrt body.
```
if x == 0.0 : 
  z = x
else : 
  # do the approximation
```

charles-cooper commented on vyper/functions/functions.py:
```diff
@@ -1224,6 +1227,61 @@ def _can_compare_with_uint256(operand):
+        new_var_pos = arg.args[0]
+    # Other locations need to be copied.
+    else:
+        new_var_pos = context.new_placeholder(x_type)
```
Maybe we can just use FREE_VAR_SPACE? Not sure if it will conflict with anything else

jacqueswww replied:
FREE_VAR_SPACE will be overwritten, that's why we allocate new memory.

charles-cooper commented on vyper/functions/functions.py:
```diff
@@ -1224,6 +1227,61 @@ def _can_compare_with_uint256(operand):
+    x_type = BaseType('decimal')
+    placeholder_copy = ['pass']
+    # Steal current position if variable is already allocated.
+    if arg.value == 'mload':
```
Does this still work if `arg.args[0]` is not a literal int?

charles-cooper replied:
I guess it should .. I have my reservations about this pattern in general though because the frontend does not check that `x` does not get overwritten. I like the idea of always using `FREE_VAR_SPACE` because it's cheap and easy to reason about :)

jacqueswww replied:
ha I know a way to fix that ;)

charles-cooper replied:
Ah, missed that one. Nice!

charles-cooper commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -0,0 +1,63 @@
+    c = get_contract_with_gas_estimation(code)
+    val = Decimal('33.33')
+
+    assert c.test(val) == decimal_sqrt(val)
```
This doesn't seem to work for values less than 1.
```python
        c = get_contract_with_gas_estimation(code)
        val = Decimal('0.999') # Decimal('33.33')
        print(c.test(val))
    
>       assert c.test(val) == decimal_sqrt(val)
E       AssertionError: assert Decimal('0.9995') == Decimal('0.9994998749')
E        +  where Decimal('0.9995') = <function gas_estimation_decorator.<locals>.decorator at 0x7f7f91e8e2f0>(Decimal('0.999'))
E        +    where <function gas_estimation_decorator.<locals>.decorator at 0x7f7f91e8e2f0> = <tests.conftest.VyperContract object at 0x7f7f91e629e8>.test
E        +  and   Decimal('0.9994998749') = decimal_sqrt(Decimal('0.999'))

tests/parser/types/numbers/test_sqrt.py:40: AssertionError
```

jacqueswww replied:
found the bug:

`if z >= y:` instead of `if z == y:`.

and will write a test accordingly ;)

charles-cooper commented on vyper/parser/memory_allocator.py:
```diff
@@ -0,0 +1,24 @@
+)
+
+
+class MemoryAllocator():
```
This class is cool :)

jacqueswww replied:
:)

charles-cooper commented:
@jacqueswww left some comments. This inline function thing is really cool!

davesque commented on vyper/parser/memory_allocator.py:
```diff
@@ -0,0 +1,24 @@
+)
+
+
+class MemoryAllocator():
```
...and another thing.  There's no need for the `()` after the class name.

jacqueswww commented:
@davesque main reason is that was the convention on repo when I joined, also I think it reads well, especially because there is a difference between comments and docs. Comments augment the code or 'narrative' of the code were reading the code alone isn't good enough, docblocks is good for understanding how to call a function. The fact that docblocks are required to be fat by convention. A good balance between docblocks and comments is what I prefer over a codebase with too much docblocks, and But that's just my completely biased  and super objective opinion.

fubuloubu commented on vyper/functions/functions.py:
```diff
@@ -1224,6 +1227,65 @@ def _can_compare_with_uint256(operand):
+    arg = args[0]
+    sqrt_code = """
+assert x >= 0.0
+z: decimal
```
So, this inline code injection is creating a variable? Would this variable be directly assigned (i.e. on every update in the inline code), or assigned at the end? `z` could be a memory or storage item.

Also, I think to be clearer, this mini-language inside our compiler should make it more explicitly clear what `x`, `y`, and `z` are, perhaps through some overhead or secondary syntax that shows they can be name-managled into working in case they shadow other variables. (it does do this, right?)

jacqueswww replied:
It's not a mini language it's actual vyper. This isn't variable shadowing, the memory stays allocated no shadowing occurs, to rest of the compiler it just continues allocating. Nothing gets mangled the compiler is completely oblivous to this, and this is just another form of what happens a ton of places already just in LLL.

fubuloubu replied:
Okay, my misunderstanding then. You know what my concern was :)

fubuloubu commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -0,0 +1,116 @@
+    b: decimal = 1.0
+    c: decimal = 2.0
+    d: decimal = 3.0
+    e: decimal = sqrt(a)
```
I want to see a test case that uses `x`, `y`, and `z` to show variable shadowing is not an issue with this injected code

jacqueswww replied:
This already proves that.

jacqueswww replied:
pushed a commit, check it out ;)

jacqueswww replied:
You did however give me an Idea, Maybe we should rollback the counter, this might give gas a improvement.

fubuloubu replied:
Nice!

fubuloubu commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -0,0 +1,116 @@
+)
+
+import pytest
+
```
Might be a good place for some `hypothesis` test cases

fubuloubu replied:
cc @davesque

fubuloubu commented:
Yassss, more hypothesis plz :smiley_cat:

jacqueswww commented:
@fubuloubu @pipermerriam Hypothesis is awesome! :rocket:  :man_dancing:  I struggled a lot until I added `getcontext().prec = 168` was very important.
@charles-cooper added CompilerPanic as requested.

I think I have now addressed/resolved all issues mentioned.

pipermerriam commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -1,22 +1,37 @@
+    SizeLimits,
+)
+
+getcontext().prec = 168
```
I'm a big fan of using the `decimal.localcontext()` context manager API as it 1) preserves existing global settings upon exit and 2) gives stronger guarantees that you are indeed working with the context you desire within the specific block of code in question.

jacqueswww replied:
Ah cool, will do :+1:

jacqueswww replied:
@pipermerriam I tried using localcontext() for theses tests now. But I think it doesn't play well with tests - I'd have to wrap each test that currently just uses `Decimal` with a with statement - which seems like serious boiler plate (for tests) e.g.
```python
def test_sqrt_variable(get_contract_with_gas_estimation):
    code = """
@public
def test(a: decimal) -> decimal:
    return sqrt(a)

@public
def test2() -> decimal:
    a: decimal = 44.001
    return sqrt(a)
    """

    c = get_contract_with_gas_estimation(code)

    with localcontext() as decimal_ctx:
        decimal_ctx.prec = PRECISION
        val = Decimal('33.33')
        assert c.test(val) == decimal_sqrt(val)

        val = Decimal('0.1')
        assert c.test(val) == decimal_sqrt(val)

        assert c.test(Decimal('0.0')) == Decimal('0.0')
        assert c.test2() == decimal_sqrt(Decimal('44.001'))

        with pytest.raises(TransactionFailed):
            c.test(Decimal('-1.1'))

```

pipermerriam commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -111,6 +129,33 @@ def test(a: decimal) -> decimal:
     assert vyper_sqrt == actual_sqrt
+
+
+class TestFuzzSqrt:
```
I think you can drop this class as I *think* hypothesis tests interoperate just fine with pytest fixtures (I may be wrong).

jacqueswww replied:
I was struggling with hitting the deadline very easily because of redeploying the code every time with get_contract, which is good for our regular tests - but bad for a fuzz test ;)

fubuloubu replied:
It does interoperate just fine with pytest fixtures, but there are sometimes performance things that I've found to be off. Hypothesis is a hog :pig:

pipermerriam replied:
Hrm, does hypothesis not respect fixture scopes?  Can you not use a `"module"` scoped fixture so that pytest only runs the deployment once?

jacqueswww replied:
I wanted to reuse the function scoped get_contract, which blocked me. But I will fix accordingly ( by making a get_contract_mofule).

jacqueswww replied:
Sorted now :)

jacqueswww commented:
@fubuloubu yes there is a assert for it, and I updated the docs to say `x > 0.0.` ;)

jacqueswww commented:
Also I tested for it: https://github.com/ethereum/vyper/pull/1351/files#diff-207daf0d816542a0095be46dc1b08da5R70

jacqueswww commented:
So interestingly enough I started adding `uint256` support for this, but somehow I can't get to model a `math.floor(math.sqrt())` or `round(0.5+math.sqrt())` equivalent setup to match to it. For instance I could model up to val = 255 and then suddenly it would pop up with 15 != 16. :laughing: I think  we should leave uint256 for another PR?

fubuloubu commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -0,0 +1,161 @@
+    assert c.test2() == decimal_sqrt(Decimal('44.001'))
+
+    with pytest.raises(TransactionFailed):
+        c.test(Decimal('-1.1'))
```
Can we change this to
```suggestion
        c.test(Decimal('-1E-10'))
```
Ensures that the "largest" possible negative number reverts (closest to zero we can test for)

fubuloubu commented on tests/parser/types/numbers/test_sqrt.py:
```diff
@@ -0,0 +1,161 @@
+
+    with pytest.raises(TransactionFailed):
+        c.test(Decimal('-1.1'))
+
```
Also do this to show the "smallest" possible negative number reverts (furthest away from zero we can test for)
```suggestion

    with pytest.raises(TransactionFailed):
        c.test(Decimal(SizeLimits.MINNUM))
```

charles-cooper replied:
Can we also test max decimal (binary representation `0xff..`)

fubuloubu replied:
P.S. there's a way to cherry pick specific cases with hypothesis

pipermerriam replied:
@fubuloubu is referring to https://hypothesis.readthedocs.io/en/latest/reproducing.html#hypothesis.example

jacqueswww replied:
@fubuloubu I added and additional test for the invalid range.

fubuloubu replied:
Perfect!

fubuloubu commented:
I honestly don't think we should allow `uint256` for `sqrt` as the result may most certainly be a decimal (and we would probably need a different algorithm)

charles-cooper commented:
> I honestly don't think we should allow `uint256` for `sqrt` as the result may most certainly be a decimal (and we would probably need a different algorithm)

I agree, the user can easily convert to decimal and back (although they need to be conscious of overflow possibility)

fubuloubu commented:
Doesn't the conversion from `uint256` to `decimal` revert if an overflow condition is present? (isn't that in the runtime clamp?)

charles-cooper commented:
> Doesn't the conversion from `uint256` to `decimal` revert if an overflow condition is present? (isn't that in the runtime clamp?)

Right so there are `uint256` numbers that you can't use convert-to-decimal-and-back to calculate the integer square root for them.

fubuloubu commented:
Oh, I missed the part where you said "_they_ need to" lol :+1:
